File: daily/2024/2940_Find_Building_Where_Alice_and_Bob_Can_Meet.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

class Solution {
public:
    vector<int> leftmostBuildingQueries(vector<int>& heights, vector<vector<int>>& queries) {
        int n = heights.size();
        int q = queries.size();
        vector<vector<pair<int,int>>> Q(n);
        vector<int> ans(q, -1);
        for (int i = 0; i < q; i++){
            int a = queries[i][0];
            int b = queries[i][1];
            if (a > b) swap(a, b);
            if (a == b || heights[a] < heights[b]) ans[i] = b;
            else Q[b].push_back({heights[a], i});
        }
        vector<pair<int,int>> monotonic;
        for (int i = n-1; i >= 0; i--){
            // int monotonic_size = monotonic.size();
            for (auto& [val, idx] : Q[i]){
                if (monotonic.empty() || monotonic[0].first <= val) ans[idx] = -1;
                else ans[idx] = monotonic[search(monotonic, val)].second;
            }
            while(!monotonic.empty() && monotonic.back().first <= heights[i]) monotonic.pop_back();
            monotonic.push_back({heights[i], i});
        }
        return ans;
    }
private:
    int search(vector<pair<int,int>>& monotonic, int val){
        int left = 0;
        int right = monotonic.size() - 1;
        int ans = -1;
        while(left <= right){
            int mid = left + (right - left) / 2;
            if (val < monotonic[mid].first){
                ans = max(ans, mid);
                left = mid + 1;
            }
            else right = mid - 1;
        }
        return ans;
    }
};

int main(){
    vector<int> heights = {6,4,8,5,2,7};
    vector<vector<int>> queries = {{0,1},{0,3},{2,4},{3,4},{2,2}};
    // vector<int> heights = {5,3,8,2,6,1,4,6};
    // vector<vector<int>> queries = {{0,7},{3,5},{5,2},{3,0},{1,6}};
    // vector<int> heights = {1,2,1,2,1,2};
    // vector<vector<int>> queries = {{0,0},{0,1},{0,2},{0,3},{0,4},{0,5},{1,0},{1,1},{1,2},{1,3},{1,4},{1,5},{2,0},{2,1},{2,2},{2,3},{2,4},{2,5},{3,0},{3,1},{3,2},{3,3},{3,4},{3,5},{4,0},{4,1},{4,2},{4,3},{4,4},{4,5},{5,0},{5,1},{5,2},{5,3},{5,4},{5,5}};
    Solution S;
    vector<int> ans = S.leftmostBuildingQueries(heights, queries);
    for (int i : ans) cout << i << " ";
    cout << endl;
    return 0;
}

File: daily_python/2940_Find_Building_Where_Alice_and_Bob_Can_Meet.py
import os
import sys
from typing import List

class Solution:
    def search(self, monotonic, val):
        left = 0
        right = len(monotonic) - 1
        ans = -1
        while left <= right:
            mid = (left + right) // 2
            if monotonic[mid][0] > val:
                ans = max(ans, mid)
                left = mid + 1
            else:
                right = mid - 1
        return ans
    
    def leftmostBuildingQueries(self, heights: List[int], queries: List[List[int]]) -> List[int]:
        n = len(heights)
        q = len(queries)
        monotonic = []
        ans = [-1] * q
        Q = [[] for _ in range(n)]
        for i in range(q):
            a = queries[i][0]
            b = queries[i][1]
            if a > b:
                a, b = b, a
            if a == b or heights[a] < heights[b]:
                ans[i] = b
                continue
            Q[b].append((heights[a], i))
        for i in range(n-1, -1, -1):
            for val, idx in Q[i]:
                if monotonic and monotonic[0][0] > val:
                    ans[idx] = monotonic[self.search(monotonic, val)][1]
            while monotonic and monotonic[-1][0] <= heights[i]:
                monotonic.pop()
            monotonic.append((heights[i], i))
        return ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            heights = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            queries = [list(map(int, i.split(','))) for i in lines[1].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append((heights, queries))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for heights, queries in dataset:
        results.append(solution.leftmostBuildingQueries(heights, queries))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')
